### PR TITLE
Add role to tally sheet mapping

### PR DESCRIPTION
**Pull request by callkalpa**
Mapping is for read, lock and unlock operations. This PR also include some checks for lock, unlock, submit, request-edit operations.

https://github.com/ECLK/results-tabulation-api/issues/279

### CODE REVIEW

File: auth/__init__.py [hunk 1]
```diff
@@ -1,22 +1,26 @@
...
 from app import db, cache
-from auth.AuthConstants import EC_LEADERSHIP_ROLE, NATIONAL_REPORT_GENERATOR_ROLE, NATIONAL_REPORT_VIEWER_ROLE, \
-    SUB, DATA_EDITOR_ROLE, POLLING_DIVISION_REPORT_VIEWER_ROLE, POLLING_DIVISION_REPORT_GENERATOR_ROLE, \
-    ELECTORAL_DISTRICT_REPORT_VIEWER_ROLE, ELECTORAL_DISTRICT_REPORT_GENERATOR_ROLE, ROLE_CLAIM_PREFIX, ADMIN_ROLE, \
-    JWT_TOKEN_HEADER_KEY
+from auth.AuthConstants import EC_LEADERSHIP_ROLE, NATIONAL_REPORT_VERIFIER_ROLE, NATIONAL_REPORT_VIEWER_ROLE, \
+    SUB, DATA_EDITOR_ROLE, POLLING_DIVISION_REPORT_VIEWER_ROLE, POLLING_DIVISION_REPORT_VERIFIER_ROLE, \
+    ELECTORAL_DISTRICT_REPORT_VIEWER_ROLE, ELECTORAL_DISTRICT_REPORT_VERIFIER_ROLE, ROLE_CLAIM_PREFIX, ADMIN_ROLE, \
+    JWT_TOKEN_HEADER_KEY, ACCESS_TYPE_READ, ACCESS_TYPE_LOCK, ACCESS_TYPE_UNLOCK
+from auth.RoleBasedAccess import role_to_read_allowed_tallysheet_types, role_to_lock_allowed_tallysheet_types, \
```
Comment (dinukadesilva) [inline auth/__init__.py:13]: From where auth.RoleBasedAccess comes? @callkalpa

Comment (callkalpa) [reply auth/__init__.py:13]: Sorry, I missed to commit that file. Pushed.

File: auth/RoleBasedAccess.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        TallySheetCodeEnum.CE_201,
+        TallySheetCodeEnum.CE_201_PV,
+        TallySheetCodeEnum.PRE_34_CO
+    ],
+    ELECTORAL_DISTRICT_REPORT_VERIFIER_ROLE: [
+        TallySheetCodeEnum.PRE_30_PD
+    ],
+    NATIONAL_REPORT_VERIFIER_ROLE: [
+        TallySheetCodeEnum.PRE_30_PD_PV,
+        TallySheetCodeEnum.PRE_30_ED
```
Comment (dinukadesilva) [inline auth/RoleBasedAccess.py]: This should include all the all island reports too

File: orm/entities/Submission/TallySheet/__init__.py [hunk 3]
```diff
@@ -42,12 +43,24 @@ def set_latest_version(self, tallySheetVersion: TallySheetVersion):
             self.submission.set_latest_version(submissionVersion=tallySheetVersion.submissionVersion)
 
     def set_locked_version(self, tallySheetVersion: TallySheetVersion):
+        if self.submittedStamp.createdBy == get_user_name():
+            raise ForbiddenException("Tally sheet submitted user is now allowed to lock/unlock.")
```
Comment (dinukadesilva) [inline orm/entities/Submission/TallySheet/__init__.py]: A small typo here "not"

File: auth/RoleBasedAccess.py (hunk 2 above)
Comment (callkalpa) [reply auth/RoleBasedAccess.py]: This is fixed.

File: orm/entities/Submission/TallySheet/__init__.py (hunk 3 above)
Comment (callkalpa) [reply orm/entities/Submission/TallySheet/__init__.py]: Fixed.

File: orm/entities/Submission/TallySheet/__init__.py [hunk 4]
```diff
@@ -42,12 +43,24 @@ def set_latest_version(self, tallySheetVersion: TallySheetVersion):
             self.submission.set_latest_version(submissionVersion=tallySheetVersion.submissionVersion)
 
     def set_locked_version(self, tallySheetVersion: TallySheetVersion):
+        if self.submittedStamp.createdBy == get_user_name():
```
Comment (dinukadesilva) [inline orm/entities/Submission/TallySheet/__init__.py]: This has to be checked only if it's data entry @callkalpa

**Review by dinukadesilva — CHANGES_REQUESTED** [review]

File: orm/entities/Submission/TallySheet/__init__.py (hunk 4 above)
Comment (callkalpa) [reply orm/entities/Submission/TallySheet/__init__.py]: Added the check.

**Review by dinukadesilva — APPROVED** [review]

**Pull request merged**